server/internal/rate/gcra_ratelimit.lua

```diff
@@ -0,0 +1,42 @@
+redis.replicate_commands()
+
+local key = KEYS[1]
+local rate = tonumber(ARGV[1])
+local cost = tonumber(ARGV[2])
+local burst_offset = tonumber(ARGV[3])
+
+-- Get current time as seconds and microseconds from Redis
+-- Convert to float for precision; adjust the epoch to be relative
+-- to Jan 1, 2020 00:00:00 GMT to avoid floating point problems.
+local jan_1_2020 = 1577836800
+local rtime = redis.call("TIME")
+local now = (rtime[1] - jan_1_2020) + (rtime[2] / 1000000)
+
+local tat = tonumber(redis.call("GET", key)) or now
```

Comment:
Could you add some comment here? Something like `-- Calculate the theoretical arrival time (TAT) of the next token`